docs/machine-learning/ml-profiling.md

```diff
@@ -0,0 +1,120 @@
+# Profiling Data Science and MLOps Code
+
+Data Science projects, especially the ones that involve Deep Learning techniques, usually are very resource intensive. One model training iteration might be multiple hours long. Although large data volumes processing genuinely takes time, minor bugs and nonoptimal implementation of some functional pieces might cause extra resources consumption.
+
+Profiling can be used to identify performance bottlenecks and see which functions are the costliest in the application code. Based on the outputs of the profiler, one can focus on largest and easiest-to-resolve inefficiencies and therefore achieve better code performance.
+Although profiling follows the same principles of any other software project, the purpose of this document is to provide profiling samples for the most common scenarios in MLOps/Data Science projects.
+
+Below are some common scenarios in MLOps/Data Science projects, along with suggestions on how to profile them.
+
+- [Profiling Data Science and MLOps Code](#profiling-data-science-and-mlops-code)
+  - [Generic Python profiling](#generic-python-profiling)
+  - [PyTorch model training profiling](#pytorch-model-training-profiling)
+  - [Azure Machine Learning pipeline profiling](#azure-machine-learning-pipeline-profiling)
+
+## Generic Python profiling
+
+Usually an MLOps/Data Science solution contains plain Python code serving different purposes (e.g., data processing) along
+with specialized model training code. Whereas many Machine Learning frameworks provide their own profiler,
+sometimes it is also very useful to profile the whole solution.
+
+There are two types of profilers: deterministic (all events are tracked) and statistical (sampling with regular intervals). In this
+document we'll cover an example of a deterministic profiler.
+
+There are many options of generic deterministic Python code profiling. The default option for profiling used to be a built-in
+[cProfile](https://docs.python.org/3/library/profile.html) profiler. Using *cProfile* one can easily profile
+either a Python script or just a chunk of code. This profiling tool produces a file that can be either
+visualized using open source tools or analyzed using `stats.Stats` class. The latter option requires setting up filtering
+and sorting parameters for better analysis experience.
+
+Below you can find an example of using *cProfile* to profile a chunk of code.
+
+```python
+import cProfile
```

Comment:
There's also the option of running it from the outside (`python -m cProfile [-o output_file] [-s sort_order] (-m module | myscript.py)`). Is this something worth considering here? Just a thought, not sure as the code is likely to be ran remotely and not by the user.

Reply:
Added, thanks a lot for the suggestion!